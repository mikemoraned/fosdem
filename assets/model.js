import { createMergeableStore } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/+esm';
import { createBroadcastChannelSynchronizer } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/synchronizers/synchronizer-broadcast-channel/+esm';
import { createLocalPersister } from 'https://cdn.jsdelivr.net/npm/tinybase@5.4.4/persisters/persister-browser/+esm';

export async function createModel() {
    const store = createMergeableStore('fosdem2025');
    
    const persister = createLocalPersister(store, 'fosdem2025');
    await persister.load();
    await persister.startAutoSave()

    const synchronizer = createBroadcastChannelSynchronizer(
        store,
        'fosdem2025SyncChannel',
        () => {
            console.log('sent a message');
        },
        () => {
            console.log('received a message');
        }
      );
    synchronizer.addStatusListener((synchronizer, status) => {
        console.log(
          `${synchronizer.getChannelName()} channel status changed to ${status}`,
        );
      });
    await synchronizer.startSync();
    return new Model(store);
}

class Model {
    constructor(store) {
        this.store = store
    }

    getBookmarkStatus(eventId) {
        return this.store.getValue(eventId) || false;
    }

    setBookmarkStatus(eventId, status) {
        console.log(`Setting bookmark status for event ${eventId} to ${status}`);
        this.store.setValue(eventId, status);
        console.log(this.store.getContent());
        console.log(this.store.getMergeableContent());
    }

    addEventListener(eventId, listenerFn) {
        this.store.addValuesListener((store, getValueChange) => {
            const [hasChanged,_oldValue,newValue] = getValueChange(eventId);
            if (hasChanged) {
                console.log(`Event ${eventId} changed to ${newValue}`);
                listenerFn(newValue);
            }
        }, false);
    }

    exportEventIdsAsText() {
        const eventIds = [];
        this.store.forEachValue((eventId, isBookmarked) => {
            if (isBookmarked) {
                eventIds.push(eventId);
            }
        });
        return eventIds.join(' ');
    }

    importEventIdsFromText(text) {
        const possibleEventIds = text.split(' ');
        possibleEventIds.forEach((possibleEventId) => {
            if (possibleEventId.match(/\d+/)) {
                this.store.setValue(possibleEventId, true);
            }
        });
    }
}