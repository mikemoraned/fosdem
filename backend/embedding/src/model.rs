use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct SubjectEmbedding {
    subject: EventArtefact,
}

impl SubjectEmbedding {
    pub fn new(subject: EventArtefact) -> SubjectEmbedding {
        SubjectEmbedding { subject }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub enum EventArtefact {
    Combined { event_id: EventId },
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct EventId(pub u32);
