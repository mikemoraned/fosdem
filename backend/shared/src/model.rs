use chrono::{Duration, NaiveDate, NaiveDateTime, NaiveTime};
use nalgebra::DVector;
use openai_dive::v1::resources::embedding::EmbeddingResponse;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, PartialEq)]
pub struct SearchItem {
    pub event: Event,
    pub distance: f64,
    pub related: Option<Vec<SearchItem>>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Event {
    pub id: u32,
    pub date: NaiveDate,
    pub start: NaiveTime,
    pub duration: u32,
    pub room: String,
    pub track: String,
    pub title: String,
    pub slug: String,
    pub url: Url,
    pub r#abstract: String,
    pub slides: Vec<Url>,
    pub presenters: Vec<Person>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Person {
    pub id: u32,
    pub name: String,
}

impl Event {
    pub fn starting_time(&self) -> NaiveDateTime {
        self.date.and_time(self.start)
    }

    pub fn ending_time(&self) -> NaiveDateTime {
        self.starting_time() + Duration::minutes(self.duration.into())
    }

    pub fn sojourner_url(&self) -> Url {
        let base_url = Url::parse("https://fosdem.sojourner.rocks/2024/event/").unwrap();
        base_url.join(&self.id.to_string()).unwrap()
    }

    pub fn nav_url(&self, current_event: &Option<Event>) -> Url {
        match current_event {
            Some(event) => Url::parse(&format!(
                "https://nav.fosdem.org/r/{}/{}",
                event.nav_room(),
                self.nav_room()
            ))
            .unwrap(),
            None => {
                let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
                location_base_url.join(&self.nav_room()).unwrap()
            }
        }
    }

    pub fn nav_room(&self) -> String {
        if self.room.contains(' ') {
            let parts: Vec<_> = self.room.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.room.to_lowercase().replace('.', "")
        }
    }
}

pub type OpenAIVector = DVector<f64>;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OpenAIEmbedding {
    pub title: String,
    pub embedding: OpenAIVector,
}

impl OpenAIEmbedding {
    pub fn embedding_from_response(response: &EmbeddingResponse) -> OpenAIVector {
        let parts = response.data[0].embedding.clone();
        OpenAIVector::from(parts)
    }
}

#[derive(Debug, Clone)]
pub struct NextEvents {
    pub now: NaiveDateTime,
    pub current: Vec<Event>,
    pub selected: Event,
    pub next: Vec<Event>,
}

#[derive(Debug)]
pub enum NextEventsContext {
    Now,
    EventId(u32),
}
