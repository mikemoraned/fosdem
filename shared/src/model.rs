use std::fmt::{Display, Formatter};
use chrono::{Duration, NaiveDate, NaiveDateTime, NaiveTime};
use nalgebra::DVector;
use openai_dive::v1::resources::embedding::{EmbeddingOutput, EmbeddingResponse};
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, PartialEq)]
pub struct SearchItem {
    pub event: Event,
    pub distance: f64,
    pub related: Option<Vec<SearchItem>>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Event {
    pub id: u32,
    pub guid: String,
    pub date: NaiveDate,
    pub start: NaiveTime,
    pub duration: u32,
    pub room: String,
    pub track: String,
    pub title: String,
    pub slug: String,
    pub url: Url,
    pub r#abstract: String,
    pub slides: Vec<Url>,
    pub presenters: Vec<Person>,
    pub links: Vec<Link>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Person {
    pub id: u32,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct Link {
    pub url: Url,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct RoomId(String); // TODO: use RoomId instead of String in Event

impl RoomId {
    pub fn new(id: String) -> RoomId {
        RoomId(id)
    }

    pub fn nav_url(&self) -> Url {
        let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
        location_base_url.join(&self.nav_room()).unwrap()
    }

    pub fn nav_room(&self) -> String {
        if self.0.contains(' ') {
            let parts: Vec<_> = self.0.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.0.to_lowercase().replace('.', "")
        }
    }
}

impl Display for RoomId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Event {
    pub fn starting_time(&self) -> NaiveDateTime {
        self.date.and_time(self.start)
    }

    pub fn ending_time(&self) -> NaiveDateTime {
        self.starting_time() + Duration::minutes(self.duration.into())
    }

    pub fn sojourner_url(&self) -> Url {
        let base_url = Url::parse("https://fosdem.sojourner.rocks/2025/event/").unwrap();
        base_url.join(&self.guid.to_string()).unwrap()
    }

    pub fn nav_url(&self, current_event: &Option<Event>) -> Url {
        match current_event {
            Some(event) => Url::parse(&format!(
                "https://nav.fosdem.org/r/{}/{}",
                event.nav_room(),
                self.nav_room()
            ))
            .unwrap(),
            None => {
                let location_base_url = Url::parse("https://nav.fosdem.org/l/").unwrap();
                location_base_url.join(&self.nav_room()).unwrap()
            }
        }
    }

    pub fn nav_room(&self) -> String {
        if self.room.contains(' ') {
            let parts: Vec<_> = self.room.split(' ').collect();
            let start = parts[0];
            start.to_lowercase().replace('.', "")
        } else {
            self.room.to_lowercase().replace('.', "")
        }
    }

    pub fn mp4_video_link(&self) -> Option<Url> {
        let video_links: Vec<Url> = self
            .links
            .iter()
            .filter(|l| l.name == "Video recording (mp4)" && l.url.to_string().ends_with(".mp4"))
            .map(|l| l.url.clone())
            .collect();
        video_links.first().cloned()
    }
}

pub type OpenAIVector = DVector<f64>;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OpenAIEmbedding {
    pub title: String,
    pub embedding: OpenAIVector,
}

impl OpenAIEmbedding {
    pub fn embedding_from_response(response: &EmbeddingResponse) -> Result<OpenAIVector, Box<dyn std::error::Error>> {
        let output = response.data[0].embedding.clone();
        match output {
            EmbeddingOutput::Float(parts) => Ok(OpenAIVector::from(parts)),
            EmbeddingOutput::Base64(base64) => {
                Err(format!("Base64 encoding not supported: {}", base64).into())
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct NextEvents {
    pub now: NaiveDateTime,
    pub current: Vec<Event>,
    pub selected: Event,
    pub next: Vec<Event>,
}

#[derive(Debug)]
pub enum NextEventsContext {
    Now,
    EventId(u32),
}
