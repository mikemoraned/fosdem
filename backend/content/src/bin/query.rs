use clap::Parser;
use content::openai::get_embedding;
use dotenvy;
use log::info;
use openai_dive::v1::api::Client;
use pgvector::Vector;
use sqlx::{postgres::PgPoolOptions, Row};
use url::Url;

/// Run a query against a remote DB
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// host address of DB
    #[arg(long)]
    host: String,

    /// query to search for
    #[arg(long)]
    query: String,

    /// how many to show
    #[arg(long)]
    limit: u8,

    /// whether to show abstract
    #[arg(long)]
    r#abstract: bool,
}

fn setup_logging_and_tracing() -> Result<(), Box<dyn std::error::Error>> {
    env_logger::init();

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    setup_logging_and_tracing()?;

    dotenvy::dotenv()?;
    let args = Args::parse();

    let password_key_name = "DB_KEY";
    let password =
        dotenvy::var(password_key_name).expect(&format!("{} is not set", password_key_name));

    info!("Connecting to DB");
    let url = format!("postgres://postgres:{}@{}/postgres", password, args.host);
    let pool = PgPoolOptions::new()
        .max_connections(5)
        .connect(&url)
        .await?;

    info!("Creating OpenAI Client");
    let api_key_name = "OPENAI_API_KEY";
    let api_key = dotenvy::var(api_key_name).expect(&format!("{} is not set", api_key_name));

    let client = Client::new(api_key);

    info!("Getting embedding for query");
    let response = get_embedding(&client, &args.query).await?;
    let embedding = Vector::from(
        response.data[0]
            .embedding
            .clone()
            .into_iter()
            .map(|f| f as f32)
            .collect::<Vec<_>>(),
    );

    let base_url = Url::parse("https://fosdem.org/2024/schedule/event/")?;

    info!("Running Query");
    let sql = "
    SELECT ev.title, ev.slug, ev.abstract, em.embedding <-> ($1) AS distance
    FROM embedding_1 em JOIN events_2 ev ON ev.title = em.title
    ORDER BY em.embedding <-> ($1) LIMIT $2;
    ";
    let rows = sqlx::query(sql)
        .bind(embedding)
        .bind(args.limit as i32)
        .fetch_all(&pool)
        .await?;
    for row in rows {
        let title: &str = row.try_get("title")?;
        let distance: f64 = row.try_get("distance")?;
        let slug: &str = row.try_get("slug")?;
        let url = base_url.join(slug)?;
        let r#abstract: &str = row.try_get("abstract")?;
        println!("title: {} (distance: {:.3})", title, distance);
        println!("url: {}", url);
        if args.r#abstract {
            println!("{}", r#abstract);
            println!();
        }
    }

    Ok(())
}
