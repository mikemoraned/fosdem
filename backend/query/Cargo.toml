[package]
name = "query"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
chrono = { workspace = true }
openai_dive = { workspace = true }
tracing = { workspace = true }
serde_json = { workspace = true }
shared = { path = "../shared" }
