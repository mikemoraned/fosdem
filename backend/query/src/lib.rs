pub mod inmemory_openai;
pub mod queryable;
