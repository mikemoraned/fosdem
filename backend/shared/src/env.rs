use tracing::info;

pub fn load_secret(name: &str) -> Result<String, Box<dyn std::error::Error>> {
    let secret = dotenvy::var(name).map_err(|e| format!("{} is not set, e: {:?}", &name, e))?;
    let suffix = secret[(secret.len() - 3)..].to_string();
    info!(
        "Loaded secret with name '{}', ending with '{}'",
        name, suffix
    );
    Ok(secret)
}

pub fn load_public(name: &str) -> Result<String, Box<dyn std::error::Error>> {
    let value = dotenvy::var(name).map_err(|e| format!("{} is not set, e: {:?}", &name, e))?;
    info!("Loaded public item with name '{}', value '{}'", name, value);
    Ok(value)
}
