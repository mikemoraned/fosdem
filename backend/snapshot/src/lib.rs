use query::{
    inmemory_openai::InMemoryOpenAIQueryable,
    queryable::{Queryable, SearchKind},
};
use serde::Serialize;
use shared::model::SearchItem;

#[derive(Debug, PartialEq, Serialize)]
pub struct Summary {
    items: Vec<ItemSummary>,
    ranks: Vec<RankSummary>,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct ItemSummary {
    event_id: u32,
    event_title: String,
}

impl ItemSummary {
    fn from_search_item(item: &SearchItem) -> ItemSummary {
        ItemSummary {
            event_id: item.event.id,
            event_title: item.event.title.clone(),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub struct RankSummary {
    event_id: u32,
    rounded_distance: f64,
    rank_in_search: usize,
}

impl RankSummary {
    fn from_ranked_search_item(ranked_item: (usize, &SearchItem)) -> RankSummary {
        let (rank, item) = ranked_item;
        RankSummary {
            rounded_distance: item.distance.clone().into(),
            rank_in_search: rank,
            event_id: item.event.id,
        }
    }
}

pub struct Snapshotter {
    queryable: InMemoryOpenAIQueryable,
}

impl Snapshotter {
    pub async fn new(
        openai_api_key: &str,
        model_dir: &std::path::Path,
    ) -> Result<Snapshotter, Box<dyn std::error::Error>> {
        let queryable = InMemoryOpenAIQueryable::connect(model_dir, openai_api_key)
            .await
            .unwrap();
        Ok(Snapshotter { queryable })
    }

    pub async fn find_related_events(
        &self,
        title: &str,
        kind: &SearchKind,
    ) -> Result<Summary, Box<dyn std::error::Error>> {
        Ok(Snapshotter::summarise(
            &self.queryable.find_related_events(title, kind, 20).await?,
        ))
    }

    fn summarise(items: &[SearchItem]) -> Summary {
        let mut item_summaries: Vec<ItemSummary> =
            items.iter().map(ItemSummary::from_search_item).collect();

        item_summaries.sort_by(|a, b| a.event_id.cmp(&b.event_id));

        let mut rank_sumamries: Vec<RankSummary> = items
            .iter()
            .enumerate()
            .map(RankSummary::from_ranked_search_item)
            .collect();

        rank_sumamries.sort_by(|a, b| a.event_id.cmp(&b.event_id));

        Summary {
            items: item_summaries,
            ranks: rank_sumamries,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::{path::PathBuf, str::FromStr};

    use shared::env::load_secret;

    use super::*;

    static TITLES: [&str; 3] = [
        "Best practices for research in open source ecosystems",
        "Where Did All the Fun Go?  And How to Bring it Back with FOSS!",
        "Staying Ahead of the Game: JavaScript Security",
    ];

    #[derive(Serialize)]
    struct TestInfo {
        model_dir: PathBuf,
        search_kind: SearchKind,
    }

    #[tokio::test]
    async fn test_find_related_events_videoonly_searchkind() {
        let kind = &SearchKind::VideoOnly;

        let openai_api_key = load_secret("OPENAI_API_KEY").unwrap();
        let model_dir = PathBuf::from_str("../shared/data/model").unwrap();
        let snapshotter = Snapshotter::new(&openai_api_key, &model_dir).await.unwrap();
        for title in TITLES {
            let similar = snapshotter.find_related_events(title, kind).await.unwrap();
            insta::with_settings!({
                info => &TestInfo {
                    model_dir: model_dir.clone(),
                    search_kind: kind.clone(),
                },
                description => title,
                omit_expression => true
            }, {
                insta::assert_yaml_snapshot!(similar);
            });
        }
    }

    #[tokio::test]
    async fn test_find_related_events_combined_searchkind() {
        let kind = &SearchKind::Combined;

        let openai_api_key = load_secret("OPENAI_API_KEY").unwrap();
        let model_dir = PathBuf::from_str("../shared/data/model").unwrap();
        let snapshotter = Snapshotter::new(&openai_api_key, &model_dir).await.unwrap();
        for title in TITLES {
            let similar = snapshotter.find_related_events(title, kind).await.unwrap();
            insta::with_settings!({
                info => &TestInfo {
                    model_dir: model_dir.clone(),
                    search_kind: kind.clone(),
                },
                description => title,
                omit_expression => true
            }, {
                insta::assert_yaml_snapshot!(similar);
            });
        }
    }
}
