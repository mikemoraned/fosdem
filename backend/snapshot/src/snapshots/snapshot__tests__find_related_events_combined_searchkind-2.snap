---
source: snapshot/src/lib.rs
description: Where Did All the Fun Go?  And How to Bring it Back with FOSS!
info: "../shared/data/model"
---
- distance: 0.48185584450340074
  rounded_distance: 0.48
  rank_in_search: 1
  event_id: 1711
  event_title: "[Ending] It's all about the email. Ugh, what?"
- distance: 0.49160805298154353
  rounded_distance: 0.49
  rank_in_search: 3
  event_id: 1776
  event_title: Journey to an open source contribution
- distance: 0.49154388523928616
  rounded_distance: 0.49
  rank_in_search: 2
  event_id: 1808
  event_title: How to Chart your own Career Path in Open Source - Panel Discussion
- distance: 0.5067876017267963
  rounded_distance: 0.51
  rank_in_search: 11
  event_id: 1830
  event_title: 20 Years of Open Source building XWiki and CryptPad
- distance: 0.4933263650810235
  rounded_distance: 0.49
  rank_in_search: 4
  event_id: 2029
  event_title: "Open Source in 2024: boundaries, burnout, business"
- distance: 0.5025012809026832
  rounded_distance: 0.5
  rank_in_search: 10
  event_id: 2305
  event_title: "Open source leadership at scale, how 1300+ people improved Drupalâ€™s multilingual features"
- distance: 0.511969246229044
  rounded_distance: 0.51
  rank_in_search: 17
  event_id: 2310
  event_title: Meritocracy or Do-ocracy - why diversity is still hard and what can we do
- distance: 0.4940026916055503
  rounded_distance: 0.49
  rank_in_search: 5
  event_id: 2328
  event_title: "Cracking the Code to Executive Support: Open Source Edition"
- distance: 0.5084979876724927
  rounded_distance: 0.51
  rank_in_search: 13
  event_id: 2741
  event_title: Take Your FOSS Project From Surviving To Thriving
- distance: 0.5146042589879563
  rounded_distance: 0.51
  rank_in_search: 19
  event_id: 3023
  event_title: Welcome to FOSDEM 2024
- distance: 0.5114524852415264
  rounded_distance: 0.51
  rank_in_search: 16
  event_id: 3113
  event_title: RHEL and CentOS and the growth of openwashing in FOSS
- distance: 0.49535694033547173
  rounded_distance: 0.5
  rank_in_search: 6
  event_id: 3163
  event_title: "Copyleft and the GPL: Finding the Path Forward to Defend our Software Right to Repair"
- distance: 0.5081329272281675
  rounded_distance: 0.51
  rank_in_search: 12
  event_id: 3254
  event_title: The State of OpenJDK
- distance: 0.5094466384407168
  rounded_distance: 0.51
  rank_in_search: 14
  event_id: 3537
  event_title: Open Source Community Updates
- distance: 0.4486985948126195
  rounded_distance: 0.45
  rank_in_search: 0
  event_id: 3544
  event_title: "Defend FOSS: From innovation to world-wide positive change"
- distance: 0.497638266054239
  rounded_distance: 0.5
  rank_in_search: 9
  event_id: 3610
  event_title: "Unpack Phabricator, Welcome Phorge - Forking the Opinionated Open Source Project Manager"
- distance: 0.49705645569927914
  rounded_distance: 0.5
  rank_in_search: 8
  event_id: 3639
  event_title: Roundtable Round-Off on FOSS CMS Collaboration
- distance: 0.513444366984524
  rounded_distance: 0.51
  rank_in_search: 18
  event_id: 3671
  event_title: "Compliance as a Community Effort: Engaging Contributors and Users"
- distance: 0.5111349836304504
  rounded_distance: 0.51
  rank_in_search: 15
  event_id: 3717
  event_title: "FOSS policy engagement: panel"
- distance: 0.4967790183625802
  rounded_distance: 0.5
  rank_in_search: 7
  event_id: 3723
  event_title: FOSDEM 2024 Highlights
